### DISCUSSION TITLE
What approach is better for communication between parent and invoked services - multicast or direct messages?

### DISCUSSION THREAD

**Discussion by ezhikov** [statelyai/xstate#1918] [Q&A]
Hi. I'm making service that invokes three more services. Some events should be shared between serves, and some don't. Right now I just have "multicast" action that send all of the events to all invoked child services, and children decide what they should do. While it works, I have a _feeling_ that more explicit messages (multiple actions with separate receivers) might be better, but can't wrap my head around "why", except that visualization sequence diagram would be much cleaner. Would like to know your opinions or experience on this matter.

**Comment by Andarist** [comment]
If you are concerned about the visualization etc then forwarding events manually to specific children definitely makes sense and is something I would reach for.

**Reply by ezhikov** [reply]
Visualization is a concern, but a low priority one. I'm thinking more about maintainability in a long run

**Reply by Andarist** [reply]
That's a hard concern to address - it feels to me that this might be a somewhat subjective matter. I don't think that any of the given options is *inherently* worse. Unless we are talking about `autoForward` - that one I would definitely try to avoid as it's, more often than not, very implicit information about the parent and child events relationship. If you are using explicit forwarding this shouldn't be as much of an issue - but I suppose it can become a refactoring hazard right now (when adding a new event to a child service without rechecking possible parents and what they forward). At least until we provide means to validate parent/child relationships.

**Reply by ezhikov** [reply]
Yes, it is definitely a subjective matter. As I said, multicast works nicely, but _feels_ like something is wrong. That is why I asked more of an opinions, rather than a recipe. Great point on validation of parent-child relationships. Will check if it will be less chatty and more verbose with one by one message forwarding
